### PR TITLE
fix(types): update return type of `Model.update`

### PR DESCRIPTION
**Pull request by guyellis**
### Pull Request Checklist

_Please make sure to review and check all of these items:_

- [x] Did you update the typescript typings accordingly (if applicable)?
- [x] Did you follow the commit message conventions explained in [CONTRIBUTING.md](https://github.com/sequelize/sequelize/blob/main/CONTRIBUTING.md)?

### Description Of Change

As per the docs in the type I've updated the type.

A few lines above the change is a description where the update() method can return 2 types. I added the 2nd type.

### CODE REVIEW

File: src/model.d.ts [hunk 1]
```diff
@@ -2207,7 +2207,7 @@ export abstract class Model<TModelAttributes extends {} = any, TCreationAttribut
         [key in keyof Attributes<M>]?: Attributes<M>[key] | Fn | Col | Literal;
     },
     options: UpdateOptions<Attributes<M>>
-  ): Promise<[number, M[]]>;
+  ): Promise<[number] | [number, M[]]>;
```
Comment (guyellis) [inline src/model.d.ts:2210]: A few lines above this the docs for this method say:

> The promise returns an array with one or two elements...

Comment (guyellis) [reply src/model.d.ts:2210]: And I can confirm that the functionality is accurate based on the docs and that the types didn't match the return type which is why I'm submitting this PR.

**Comment by guyellis** [comment]
Failing test looks like a random issue unrelated to this PR:

```
  1 failing

  1) [SQLITE] Transaction
       fails with SQLITE_BUSY when retry.match is changed:
     AssertionError: expected promise to be rejected with an error including 'SQLITE_BUSY: database is locked' but it was fulfilled with [ undefined, undefined ]
      at assertIfNotNegated (node_modules/chai-as-promised/lib/chai-as-promised.js:68:19)
      at /home/runner/work/sequelize/sequelize/node_modules/chai-as-promised/lib/chai-as-promised.js:183:17
      at async Context.<anonymous> (test/integration/transaction.test.js:739:9)
```

File: src/model.d.ts (hunk 1 above)
Comment (ephys) [inline src/model.d.ts:2210]: Suggestion to follow the format that's used in https://github.com/sequelize/sequelize/pull/14145

```suggestion
  ): Promise<[affectedCount: number, affectedRows?: M[]]>;
```

**Review by ephys — CHANGES_REQUESTED** [review]
I think we could handle this like `Model.count`: have two method signatures:
- one with option.returning set to true that returns `[affectedCount: number, affectedRows: M[]]`
- one without option.returning (or set to false) that returns `[affectedCount: number]`

**Comment by ephys** [comment]
Also you're right, the sqlite issue is something that started happening randomly recently but unrelated to your PR

**Comment by guyellis** [comment]
> I think we could handle this like `Model.count`: have two method signatures:
> 
> * one with option.returning set to true that returns `[affectedCount: number, affectedRows: M[]]`
> * one without option.returning (or set to false) that returns `[affectedCount: number]`

That was a great idea. Thank you! I've made the changes and pushed them. LMK what you think. Thanks.

File: src/model.d.ts [hunk 2]
```diff
@@ -2202,12 +2202,25 @@ export abstract class Model<TModelAttributes extends {} = any, TCreationAttribut
    * affected rows (only supported in postgres and mssql with `options.returning` true.)
    */
   public static update<M extends Model>(
+    this: ModelStatic<M>,
+    values: {
+        [key in keyof Attributes<M>]?: Attributes<M>[key] | Fn | Col | Literal;
+    },
+    options: UpdateOptions<Attributes<M>> & { returning: true }
+  ): Promise<[affectedCount: number, affectedRows?: M[]]>;
```
Comment (ephys) [inline src/model.d.ts:2210]: I think as you split the two signatures, the new signature should be:

```suggestion
  ): Promise<[affectedCount: number, affectedRows: M[]]>;
```

**Review by ephys — APPROVED** [review]
Thanks, I think we're good to go :)
Merging another PR first then I'll update & merge this one

**Comment by guyellis** [comment]
> Thanks, I think we're good to go :) Merging [another PR](https://github.com/sequelize/sequelize/pull/14151) first then I'll update & merge this one

Thank you for doing this! You are awesome! :trophy:

**Comment by Kessa713** [comment]
There's an issue with this in reference to the documentation and current functionality for `returning: (keyof TAttributes)[]` as listed [here](https://sequelize.org/v6/class/src/model.js~Model.html#static-method-update)
This breaks existing type checks when updating to v6.17.0

**Comment by ruiaraujolink** [comment]
Hi, this PR create a problem, when I pass an array of attributes this happens
Screenshot 2022-03-04 at 13 58 40

**Comment by WikiRik** [comment]
> Hi, this PR create a problem, when I pass an array of attributes this happens
> Screenshot 2022-03-04 at 13 58 40
>  

Hi! This should be fixed with #14184 right? That PR will be reviewed soon and then we will release a new version

**Comment by ruiaraujolink** [comment]
> > Hi, this PR create a problem, when I pass an array of attributes this happens
> > Screenshot 2022-03-04 at 13 58 40
> 
> Hi! This should be fixed with #14184 right? That PR will be reviewed soon and then we will release a new version

I think so, thanks.

**Pull request merged**